smartsim/_core/schemas/utils.py

```diff
@@ -1,22 +1,46 @@
-import json
 import typing as t
+from dataclasses import dataclass
 
 import pydantic
 
-_KeyT = t.TypeVar("_KeyT")
 _SchemaT = t.TypeVar("_SchemaT", bound=pydantic.BaseModel)
 
 
-class SchemaSerializer(t.Generic[_KeyT, _SchemaT]):
+@t.final
+@dataclass(frozen=True)
+class _Message(t.Generic[_SchemaT]):
+    header: str
```

Comment:
I think we might benefit from some `Field` definitions here that enforce non-zero length content for all properties.

It could maybe also remove the questionable validation and "default delimiter" being held in the registry, too.

Maybe some tests to validate wire format...

Reply:
does the header need to be non-empty? I see no problem with doing
```py
@registry.register('')  # <-- Admittedly this will make debugging hell,
                        #     but it shouldn't break anything... I guess that alone is
                        #     enough of a reason to enforce the constraint ¯\_(ツ)_/¯
class MySchema(pydantic.BaseModel):
    ...
```
I'll the restriction for both! I guess it is easier to add options rather than take them away down the line!!

Reply:
register is taking a "mapping key" which is technically unrelated to the schema, so I feel ok having separate validation. I'm with you

Reply:
For the "default delimiter", the registry needs to know what it the delimiter is for the messages that it is responsible for formatting/hydrating because it needs to be able to parse out the type information in the header in order to index its `_map`. I can add default to the `_Message` class, but a the Schema registry will still need to always supply its own, and because `_Message` is scope to this module it will never be used.

I think this conflict is leading to the "questionable validation" you mentioned, but if you could elaborate a little more I should be able remedy/justify!

Reply:
I think this is where that two-way style registry came into the prior convo. We have:

```py
def register(key, type):
  self._map[key] = type  
```

but could add:

```py
def register(key, type):
  self._forward[key] = type
  self._backward[type] = key
```

I think that might let us avoid having to "find" the mapping key in the message within the registry?

Reply:
How is `backward` different than `reverse_map` in `_to_message`? All that would do is save some iteration cycles when calling, `_to_message`, but given that we are dealing with relatively small dicts, I'm not sure I want to add the complexity over head of making sure `_forward` and `_backward` stay "in sync".

To clarify my previous comment, the delimiter is not needed to index into `_map` (or indeed `_forward`). It is need to parse out the type information the `str_` parameter of `from_string` in order to know what to index _with_.

Reply:
For reference, [this is the offending line](https://github.com/al-rigazzi/SmartSim/pull/8/files#diff-eb87d606b148ad7705ae9e182b6f5e555ace51718f6a2dce191994ca9f8213e2R68)